### PR TITLE
refactor K8sManifest

### PR DESCRIPTION
**Pull request by nimakaviani**
- differentiate between K8sObjectManifest and K8sCredentialManifest but
have both of them inherit from a higher level K8sManifest
- for each resolver, use its local call to retrieve K8s resources. this
allows custom emission of events where needed, particularly as deployment
of resources progresses. see the example case for when a K8s resource is
deployed, which cause of an oversight was not working as expected
- add tests to ensure we emit deploying, and deployed events for k8s resources

### CODE REVIEW

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+data class K8sCredentialManifest(
+    override var apiVersion: String?,
+    override var kind: String?,
+    @get:ExcludedFromDiff
+    override val metadata: Map<String, Any?>,
+    override val spec: K8sSpec?,
+    override val data: K8sData? = null) : K8sManifest(apiVersion, kind, metadata, spec, data) {
+    override fun name(): String = "${metadata[TYPE]}-${data?.account}"
```
Comment (nabuskey) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:74]: If we are missing the type field in the metadata Map, it's going to return null. Although it doesn't throw exception here, the name is going to be like `null-account1`. Wonder if we should check and make sure the type field is included in the submitted spec.

**Review by nabuskey — APPROVED** [review]
I think we should add our own annotation to resources we create. Not in this PR but in the future. It'll make garbage collection easier too.

**Comment by nimakaviani** [comment]
great idea with adding our own annotations to resources the plugin creates. added #42 to track it.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt (hunk 1 above)
Comment (nimakaviani) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:74]: yeah doing the validation when it gets submitted is probably the right thing to do.

**Pull request merged**